### PR TITLE
Hedera Schedule Service System Contract

### PR DESCRIPTION
**Pull request by Nana-EC**
**Description**:
This proposal addresses the feature gap of a smart contracts ability to issue scheduled transactions via the HAPI scheduled transactions.

Since smart contracts executions do not utilize the Hedera signature map they are unable to carry along the authorizations that the Hedera ledger uses to confirm an accounts participation and acknowledgment in a transaction. 

To address this Smart Contracts could utilize the Hedera Schedule Service by submitting a scheduled transaction to which accounts can sign / authorize as an acceptance of the desired transaction. This flow provides as easy route for asynchronous coordination of transaction approval.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+discussions-to: 
+---
+
+## Abstract
+
+This proposal addresses the feature gap of a smart contracts ability to issue scheduled transactions via the HAPI scheduled transactions.
+
+Since smart contracts executions do not utilize the Hedera signature map they are unable to carry along the authorizations that the Hedera ledger uses to confirm an accounts participation and acknowledgment in a transaction. 
+
+To address this Smart Contracts could utilize the Hedera Schedule Service by submitting a scheduled transaction to which accounts can sign / authorize as an acceptance of the desired transaction. This flow provides as easy route for asynchronous coordination of transaction approval.
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:20]:
```suggestion
To address this Smart Contracts could utilize the Hedera Schedule Service by submitting a scheduled transaction to which accounts can sign / authorize as an acceptance of the desired transaction. This flow provides as easy route for asynchronous coordination of transaction approval.
```
To address this, Smart Contracts could utilize the Hedera Schedule Service by submitting a scheduled transaction to which accounts can subsequently sign / authorize as an acceptance of the desired transaction. This flow provides an easy route for asynchronous coordination of transaction approval.

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+## Motivation
+
+In many decentralized scenarios a contract may issue a transaction that would require participation by multiple entities.
+
+However, under the Hedera Smart Contract Service (HSCS) Security Model v2 it is not possible to authorize a contract in advance to modify an accounts property or cause a debit to their balance without their authorization. This essentially means multi party operations made challenging if not infeasible on smart contracts.
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:26]:
```suggestion
This essentially means multi party operations are made challenging if not infeasible when using smart contracts.
```

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+This is a step back in the Hedera UX that was made easier by the use of in transaction signatures.
+
+## Rationale
+
+By providing a secure mechanism to acquire asynchronous authorization from multiple accounts, smart contracts can continue to be used for more decentralized operations whiles still maintaining the integrity of account sovereignty by allowing accounts to approve confirm their participation in a transaction.
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:32]:
```suggestion
By providing a secure mechanism to acquire asynchronous authorization from multiple accounts, smart contracts can continue to be used for more decentralized operations while still maintaining the integrity of account sovereignty by allowing accounts to approve and confirm their participation in a transaction.
```

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+## User stories
+
+1. As an EOA I would like to initiate a smart contract transaction that schedules a supported transaction.
+2. As an EOA I would like to initiate a smart contract transaction that allows me to sign a scheduled transaction.
+3. As an EOA I would like to initiate a smart contract transaction that prompts a contract to authorize a scheduled transaction.
+4. As an EOA I would like to initiate a smart contract transaction that allows me to extract information about a scheduled transaction.
+  
+## Specification
+
+The ledger HSCS will utilize the existing scheduled transaction service supported on the ledger within the system contract logic.
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:43]:
```suggestion
HSCS will utilize the existing scheduled transaction service supported on the ledger within the system contract logic. 
```

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+|---------------|---------------------------------------------------------------------------------------------------|-----------------------|
+| `0xf0637961`  | `authorizeSchedule(address) external returns (int64 responseCode)`                                | Y                     |
+| `0x5e147101`  | `getScheduledTransactionInfo(address) external returns (bytes memory transactionProtobufBytes)`   | Y                     |
+| `0xd797b304`  | `signSchedule(address) external returns (int64 responseCode)`                                     | N                     |
+| `0x358eeb03`  | `signSchedule(address, bytes) external returns (int64 responseCode)`                              | Y                     |
+
+Since `signSchedule(address scheduleAddress) returns (int64 responseCode)` relies on an implicit signature it will only callable by EOA’s via the IHRC facade.
+In this case the signature will be the inner ECDSA signature found in the RLP encoded `EthereumTransaction`. 
+For `Contract` and `ContractCreate`  any applicable signature found in the signature map will be utilized as in `ScheduleSign`.
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:60]: I'm not sure if I understand what this sentence means?  What is `ScheduleSign` referencing?

File: HIP/hip-0000-schedule-service-system-contract.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+No further protobuf or application level changes are needed as HSS is already implemented and functional.
+
+## Backwards Compatibility
+
+Backwards compatibility is ensured as no existing features are modified. Similar to HTS system contract this HIP simply exposes HAPI entity functionality and the system comtract will utilize the same HAPI services on the node.
+
+## Security Implications
+
+Existing security consideration such as throttles will remain applicable. 
+Additional considerations may include
```
Comment (lukelee-sl) [inline HIP/hip-0000-schedule-service-system-contract.md:72]: Finish or remove.

Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: The rest of the section are the completion of the section. Let me update it to be more explicit

File: HIP/hip-0000-schedule-service-system-contract.md (hunk 1 above)
Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: Updated

File: HIP/hip-0000-schedule-service-system-contract.md (hunk 2 above)
Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: Updated

File: HIP/hip-0000-schedule-service-system-contract.md (hunk 3 above)
Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: Updated

File: HIP/hip-0000-schedule-service-system-contract.md (hunk 4 above)
Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: Updated

File: HIP/hip-0000-schedule-service-system-contract.md (hunk 6 above)
Comment (Nana-EC) [reply HIP/hip-0000-schedule-service-system-contract.md]: Updated

**Review by mgarbs — APPROVED** [review]

**Pull request merged**